### ISSUE TITLE
Generation of list of fragments in colBlockApply very slow for large objects

### ISSUE THREAD

**Issue by wmacnair** [tatami-inc/beachmat#20]
This issue is related to the function `sumCountsAcrossCells` in `scater`/`scuttle`, so I wasn't quite sure where to raise it - happy to put it there instead.

I've been hacking together a parallelized version of `aggregateData` in the `muscat` package, via `sumCountsAcrossCells`, and applying it to a large `sce` object (multiple 100k of cells). With the version of `scater` from a few months ago, it took ~40 minutes with 16 cores, and since some recent update it takes more like 4 hours I think. Almost all of this extra time is spent in lines 132:151 in `colBlockApply.R`.

I think the difference is in how the blocks are split, possibly due to the change in this `scuttle` commit:
https://github.com/LTLA/scuttle/commit/46fd4b303c5f63dac358b361c5871e5d2a2736f3#
So it seems that `scuttle:::.splitColsByWorkers` is a lot faster than the equivalent lines in `beachmat::colBlockApply`. 

But maybe the `beachmat` version is more efficient in memory terms? I don't understand enough about parallelization to guess. Would be nice if there was some way of making the `beachmat` fragment generation code quicker, perhaps by parallelizing it.

Thanks!
Will

**Comment by LTLA** [comment]
Hard to say without a MRE. Those lines you've pointed out don't do anything except subset a matrix, provided `nopar=FALSE`. The extra operations around the subsetting are basically negligible in cost:

```r
library(Matrix)
mat <- rsparsematrix(20000, 100000, density=0.001) # 20k genes, 100k cells.
system.time(out <- colBlockApply(mat, colSums, BPPARAM=MulticoreParam(10)))
##    user  system elapsed 
##   2.297   0.532   1.388 
```

So, the subsetting loop has pretty low overhead. (The actual subsetting is the same cost in both the old and new functions.) I suggest you `debug(beachmat:::.blockApply2)` and check out what's going on, in particular the length of `fragments`.

**Comment by wmacnair** [comment]
Ok, thanks. 

I have 1700 fragments.

Yes, all those lines do is subset a matrix, but that subsetting is very slow, i.e. the call 
```r
block <- do.call("[", c(common.args, idx))
```
is where `beachmat:::.blockApply2` spends most of its time (it's something like 12s per call, and with 1700 fragments this explains the multiple hours necessary).

I've had a look and my matrix is more like
```r
mat <- rsparsematrix(30000, 600000, density=0.05)
```
which is big enough that it takes a long time even to generate it.

So perhaps it's expected that subsetting something so massive takes a long time, but what's puzzling is the difference in time between the previous and current iterations. Any guesses on what that might come from?

**Comment by LTLA** [comment]
In the past, the number of subset operations was determined by the number of assigned workers; so, for example, there would only be 16 subset operations to split the matrix into 16 chunks, each of which is distributed to a single worker.

However, this is not technically safe as the user-supplied function may do arbitrarily memory-consuming operations, e.g., convert the matrix into a dense format, at which point your computer will probably brick. The latest version of **beachmat** prevents this from happening by constraining each chunk to be of size no greater than `DelayedArray::getAutoBlockSize()`. The default is 100 MB, so any copies or loss of sparsity, etc. will only cause relatively small reallocations that will be tolerated by the R session.

The cost of this constraint is that you need a lot more chunks and a lot more subsetting operations. The best solution is to increase the acceptable block size, e.g., to 1 GB with `setAutoBlockSize(1e9)`; this will cut down the number of subsetting operations by 10-fold. However, this assumes that you actually have enough memory to support an increased block size for every analysis.

If you want to parallelize the subsetting itself, this is achieved by wrapping the input matrix in a `DelayedArray`. The subset operation is delayed until the matrix values are to be extracted in each worker. While appealing, this is not the default approach as it requires the entire matrix to be serialized to each worker as well, which is unpalatable for non-`MulticoreParam` backends.

Having said that, I guess I could detect whether a multicore backend is being used and switch to a distributed subsetting model. I suppose I could also trust that the caller will not supply a memory-busting function and make more aggressive chunks.

**Comment by wmacnair** [comment]
Ok, so we have the constraint of trying to make things memory-safe trading off against speed.

One thing I don't totally get is the `makeNindexFromArrayViewport` call in the context of `scuttle:::.summarize_assay`. `makeNindexFromArrayViewport` doesn't use any values of `by.group` set in `scuttle:::.summarize_assay`, right? Could it be more efficient to have a special chunker within `scuttle` that chunks based on `by.group`? Splitting into subchunks if necessary to meet the block size requirement. But perhaps that would require extensive replumbing.

(Ah I've just understood that the chunking is done across rows. But given that you're doing chunking anyway, and chunking appears to be the rate-limiting step, it would seem like you could gain some efficiency by switching to column-chunking.)

**Comment by LTLA** [comment]
Chunking by row is generally the better bet as there is no guarantee that there will be enough groups of similar size to keep all workers occupied, while there are usually enough genes that you can distribute them pretty well across a large number of workers. One could chunk across cells but this introduces problems with numerical precision from differences in the sequence of addition operations used to reduce the result, meaning that your results will change slightly depending on the number of workers.

The real problem is probably the fact that row subsetting of `*gCMatrix`es is painfully slow for large matrices, with a high constant overhead per subset operation due to the column-focused nature of the data structure. This can be resolved at the **beachmat** level with some TLC, though I would have to find some time to do so. In any case, I would prefer not to implement **scuttle**-specific workarounds, I don't want to have to re-solve the same problem in each of 5-10 affected packages.

I suggest you verify that the _number_ of subsetting operations is indeed the cause of the slowdown, e.g., by seeing if you can speed it up by bumping up `setAutoBlockSize()`. This will make it easier for me to prioritize.

**Comment by wmacnair** [comment]
I set `setAutoBlockSize(1e9)`, and now each subsetting call takes ~20s. So it does seem to be the number of operations that's limiting.

I'll work out some (non-memory safe) hack of my own for now. Thanks for the explanations.

**Comment by LTLA** [comment]
Try the `fastsparse` branch and see whether that does any better. For me, running the below:

```r
library(Matrix)
y <- rsparsematrix(20000, 50000, density=0.05)

library(BiocParallel)
library(beachmat)
system.time(out <- rowBlockApply(y, rowSums, BPPARAM=MulticoreParam(10)))
```

Takes 15 seconds on `master` and 1.5 seconds with `fastsparse`.

**Comment by wmacnair** [comment]
Oh nice, thanks will give that a try.

**Comment by LTLA** [comment]
So... good? Bad? No difference? I'm going to assume it was an improvement unless told otherwise.

**Comment by LTLA** [comment]
I'm going to assume it all worked out for the best. Consider these timings on `master`:

```r
# Splitting using the fast method
chunk_by_row_fast <- function(x, grid) {
    grid <- beachmat:::.prepare_sparse_row_subset(x, grid)
    lapply(grid, FUN=beachmat:::.subset_matrix, x=x, vp=NULL)
}

# Old-school splitting with Matrix's [ operator
chunk_by_row_ref <- function(x, grid) {
    lapply(seq_along(grid), FUN=function(i) beachmat:::.subset_matrix(x, grid[[i]]))
}

y <- Matrix::rsparsematrix(20000, 100000, density=0.01)
grid <- rowAutoGrid(y)
system.time(out <- chunk_by_row_ref(y, grid))
##    user  system elapsed 
##  12.788   0.128  12.923 
system.time(alt <- chunk_by_row_fast(y, grid))
##    user  system elapsed 
##   0.906   0.056   0.961 

identical(out, alt)
## [1] TRUE
```

Available in **beachmat** 2.7.4. If all goes well I will migrate this to release.

**Comment by wmacnair** [comment]
Sorry, haven't had a chance to try it out. But it certainly looks like a solution.

**Comment by wmacnair** [comment]
Hello again, I've been using this useful function again, hence resurrecting this discussion.

I've been having similar issues to before, namely that aggregating from an `sce` to a pseudobulk with `summarizeAssayByGroup` is slower than it could be. Last time I was here I wrote that to aggregate my `sce`, a function in an old version of `scater` was taking ~40 minutes with 16 cores. I'm now running `muscat::aggregateData`, which has `summarizeAssayByGroup` under the hood, and with 8 cores it's taken 3+ hours so far to get to 38%.

I have to confess, I still don't totally get the discussion above. 

However, I've noted a few things, and maybe they will give pointers to you:
- As I mentioned, the current implementation of `summarizeAssayByGroup` is still slower than some code in an old version of `scater`. I haven't been able to track down exactly which function that was; I guess it was `sumCountsAcrossCells` in a version `scater`/`scuttle` from < Dec 2020.
- You chunk by row, but then sum across columns. Doesn't the storage of sparse matrices guarantee that one of these operations will be slow? It would feel more sensible to chunk by column throughout, no?
- You raise the issue that users might request a memory-exploding function, which is a valid concern. However, in practice, I imagine that 90%+ of calls to the function will request one of a small number of functions (including `sum`, `mean`, `prop.detected`). What about checking whether the requested function is in some sensible list, then implementing the faster, less memory safe chunking for any functions in that list?
- I'm a bit surprised by the memory usage. My `sce` object is 19GB when saved (uncompressed), but with 8 cores, the job is currently using 240GB. Is that surprising to you? I'm using `MulticoreParam` with 8 cores.

I hope these make sense!

Thanks
Will

**Comment by wmacnair** [comment]
A further comment:

I thought I would try a `data.table` version. Running `summarizeAssayByGroup` with 8 cores takes multiple hours. The following code ran in less than 4 minutes with 4 cores, and used less memory:
```
t_start   = Sys.time()
mat_dt    = data.table(
  i         = counts(sce)@i + 1,
  j         = counts(sce)@j + 1,
  count     = counts(sce)@x
  ) %>% 
  .[, cell_id := colnames(sce)[j] ] %>% 
  setkey("cell_id")
t_stop    = Sys.time()
message('time to create dt: ', round(difftime(t_stop, t_start, units = 'secs')), ' seconds')

clusts_dt = clusts_dt %>% setkey("cell_id")
t_start   = Sys.time()
pb_dt     = mat_dt %>% merge(clusts_dt, by = "cell_id") %>% 
  .[, .(count = sum(count), .N), by = .(i, cluster, sample_id) ]
t_stop    = Sys.time()
message('time to aggregate: ', round(difftime(t_stop, t_start, units = 'secs')), ' seconds')
```

I think I will be moving to this approach for the moment...!
Will

**Comment by LTLA** [comment]
I don't remember much from this discussion, but I also didn't change the code since, so I'm not sure what happened.

I'll mention that, for most of my applications, I use `scran.chan::aggregateAcrossCells.chan`, which is pretty fast:

```r
library(scRNAseq)
sce <- BhaduriOrganoidData() # 240k cells

library(scran.chan)
mat <- initializeSparseMatrix(assay(sce))
system.time(res <- aggregateAcrossCells.chan(mat, list(cloneid=sce$cloneid, donorid=sce$donorid)))
##    user  system elapsed 
##   0.761   0.000   0.761 
```

Not nearly as large as your dataset, but hopefully it should still be faster than `aggregateAcrossCells`.

Anyway, as for your questions:

> * You chunk by row, but then sum across columns. Doesn't the storage of sparse matrices guarantee that one of these operations will be slow? It would feel more sensible to chunk by column throughout, no?

Yes, it would be, but I can't chunk by column and parallelize in a manner that doesn't change the result for floating-point matrices. I'd like for the output to be bit-identical regardless of the number of cores and size of the chunks, and for floating-point summation, this will not be the case when the chunks determine the sequence of addition operations. Bit-level equality of results is, unfortunately, important as algorithms like UMAP and t-SNE are highly sensitive to differences.

> * You raise the issue that users might request a memory-exploding function, which is a valid concern. However, in practice, I imagine that 90%+ of calls to the function will request one of a small number of functions (including `sum`, `mean`, `prop.detected`). What about checking whether the requested function is in some sensible list, then implementing the faster, less memory safe chunking for any functions in that list?

Not sure what you mean specifically, but more generally, yes, we could implement a fast-path for `dgCMatrix` objects. That would probably be a reasonable idea. Don't bother parallelizing, just scan through the matrix and process it all.

I might work on this if I have time. Or you could make a PR if you have a good idea of what I mean. Probably would help speed up a few packages dependent on `summarizeAcrossGroups`. Maybe @HelenaLC might be interested in helping.

> * I'm a bit surprised by the memory usage. My `sce` object is 19GB when saved (uncompressed), but with 8 cores, the job is currently using 240GB. Is that surprising to you? I'm using `MulticoreParam` with 8 cores.

This is unfortunate but unsurprising. I did some work on this a long time ago; from what I remember R's parallelization messes with the garbage collector, because each child process thinks it has all the allocated memory to work with, whereas the "real" space per process is actually the total available space divided by the number of processes. As such, the GC doesn't get triggered as much as it should, causing memory usage to grow by a lot more than it might otherwise. I recall making a patch to R to try to get this under control, but the powers that be were unconvinced. Oh well.

These days I parallelize at the C++ level, so I don't have to try to tiptoe around R's memory management.

**Comment by LTLA** [comment]
FWIW the GitHub (and soon BioC-devel) version of **scuttle** (1.9.4) now has a dedicated fast path for sparse matrices. 

This should be pretty fast.

```r
library(scRNAseq)
sce <- BhaduriOrganoidData() # 240k cells
system.time(summarizeAssayByGroup(assay(sce), ids=colData(sce)[,c("cloneid", "donorid")], statistics="sum"))
##    user  system elapsed 
##   0.545   0.000   0.545 
```

**Comment by wmacnair** [comment]
Hi

Thanks for the pointer to `scran.chan`, I wasn't aware of that. I'll try it out. And for the pointer to the faster `scuttle`.

That's interesting regarding the bit-level differences. I wasn't aware that that could be an issue.

My current fix uses `data.table`, and I guess you wouldn't want to include an additional package dependency, so likely not a good solution.

Thanks for the context on R's parallelization. It does seem like R's (or maybe `BiocParallel`'s?) approach to parallel processing could be improved, at least judging from the issues I have on our cluster...

You parallelize with Cpp - just out of curiosity, have you played with Rust?

**Comment by LTLA** [comment]
> My current fix uses `data.table`, and I guess you wouldn't want to include an additional package dependency, so likely not a good solution.

Correct. I've been repeatedly burnt by **data.table**'s mutability, so I'm not going back there again.

> It does seem like R's (or maybe `BiocParallel`'s?) approach to parallel processing could be improved, at least judging from the issues I have on our cluster...

Any intervention would probably have to occur at the R level, given that we'd be modifying the behavior of the garbage collector. **BiocParallel** already attempts to fix this issue (Bioconductor/BiocParallel#124) - clearly it's not enough.

> You parallelize with Cpp - just out of curiosity, have you played with Rust?

No, I didn't see a strong use case for my applications. Well, other than being new and fancy.
